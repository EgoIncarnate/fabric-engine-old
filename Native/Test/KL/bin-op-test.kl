struct Vec3
{
  Scalar x;
  Scalar y;
  Scalar z;
};

function Vec3( Scalar x, Scalar y, Scalar z )
{
  self.x = x;
  self.y = y;
  self.z = z;
}

function Vec3 +( Vec3 lhs, Vec3 rhs )
{
  var Vec3 result;
  result.x = lhs.x + rhs.x;
  result.y = lhs.y + rhs.y;
  result.z = lhs.z + rhs.z;
  return result;
}

function entry()
{
  var Vec3 lhs(1.0,2.0,3.0), rhs(4.0,5.0,6.0);
  report lhs + rhs;
}
