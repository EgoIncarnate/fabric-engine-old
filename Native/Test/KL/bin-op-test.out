GlobalList
  StructDecl( 'Vec3' )
    StructDeclMember( 'x', Scalar )
    StructDeclMember( 'y', Scalar )
    StructDeclMember( 'z', Scalar )
  Function( '' ('__construct_Vec3__Scalar__Scalar__Scalar'), void )
    ParamList
      Param( 'self', Vec3, l-value )
      Param( 'x', Scalar, r-value )
      Param( 'y', Scalar, r-value )
      Param( 'z', Scalar, r-value )
    CompoundStatement
      StatementList
        ExprStatement
          AssignOp(ASSIGN_OP)
            StructMemberOp( 'x' )
              Var('self')
            Var('x')
        ExprStatement
          AssignOp(ASSIGN_OP)
            StructMemberOp( 'y' )
              Var('self')
            Var('y')
        ExprStatement
          AssignOp(ASSIGN_OP)
            StructMemberOp( 'z' )
              Var('self')
            Var('z')
  Function( '' ('__operator_ADD__Vec3__Vec3'), Vec3:r-value )
    ParamList
      Param( 'lhs', Vec3, r-value )
      Param( 'rhs', Vec3, r-value )
    CompoundStatement
      StatementList
        StatementList
          VarDecl( 'result', Vec3 )
        ExprStatement
          AssignOp(ASSIGN_OP)
            StructMemberOp( 'x' )
              Var('result')
            BinOp( ADD )
              StructMemberOp( 'x' )
                Var('lhs')
              StructMemberOp( 'x' )
                Var('rhs')
        ExprStatement
          AssignOp(ASSIGN_OP)
            StructMemberOp( 'y' )
              Var('result')
            BinOp( ADD )
              StructMemberOp( 'y' )
                Var('lhs')
              StructMemberOp( 'y' )
                Var('rhs')
        ExprStatement
          AssignOp(ASSIGN_OP)
            StructMemberOp( 'z' )
              Var('result')
            BinOp( ADD )
              StructMemberOp( 'z' )
                Var('lhs')
              StructMemberOp( 'z' )
                Var('rhs')
        ReturnStatement
          Var('result')
  Function( 'entry' ('entry'), void )
    ParamList
    CompoundStatement
      StatementList
        StatementList
          InitializedVarDecl( VarDecl( 'lhs', Vec3 ), ArgList )
            ArgList
              ConstScalar( 1 )
              ConstScalar( 2 )
              ConstScalar( 3 )
          InitializedVarDecl( VarDecl( 'rhs', Vec3 ), ArgList )
            ArgList
              ConstScalar( 4 )
              ConstScalar( 5 )
              ConstScalar( 6 )
        Report
          BinOp( ADD )
            Var('lhs')
            Var('rhs')
{x:5,y:7,z:9}
